### DISCUSSION TITLE
How to only allow and format certain characters for `TextField` input.

### DISCUSSION THREAD

**Discussion by joshuakcockrell** [skiptools/skip#144] [Q&A] [answered]
This pattern comes up a lot in my app with user input:

- User is shown a `TextField`
- Don't allow users to type anything other than `0123456789`
- Don't allow users to copy paste dumb text like `Robert'); DROP TABLE Students;--` (Yes, I also have backend string sanitization)

What I've tried:

`CharacterSet`
<img width="690" alt="Screenshot 2024-06-12 at 12 15 56 PM" src="https://github.com/skiptools/skip/assets/8979400/43a4dad3-15fe-46c6-bacb-c9976ac11430">

`Set`
<img width="691" alt="Screenshot 2024-06-12 at 12 17 50 PM" src="https://github.com/skiptools/skip/assets/8979400/1c3b4ad2-f7ce-4324-8534-c8488c9a5301">

(it gets turned into this Kotlin)
<img width="662" alt="Screenshot 2024-06-12 at 12 18 04 PM" src="https://github.com/skiptools/skip/assets/8979400/57978b47-0708-4915-ac28-eff40c99d604">

`Set(Array(".."))`
<img width="686" alt="Screenshot 2024-06-12 at 12 18 30 PM" src="https://github.com/skiptools/skip/assets/8979400/f51958aa-3410-48f2-81b3-ef7455fc489d">

Also, while we're here.. `textContentType(..)` allows iOS to auto suggest input for the user (like if they got texted a one-time code it would auto fill, or it could suggest their address, etc.) so that would be nice to support.

<img width="587" alt="Screenshot 2024-06-12 at 12 22 54 PM" src="https://github.com/skiptools/skip/assets/8979400/55968a5b-f2f7-478c-b405-017e600a2e6f">

Thanks!

**Comment by joshuakcockrell** [comment]
[This looks like the Kotlin I want](https://stackoverflow.com/a/42232882/4975772)

```
val validChars = "0123456789".toSet()
val onlyValid = input { validChars.contains(it) }
```

I can report the following works great:

<img width="371" alt="Screenshot 2024-06-12 at 2 07 55 PM" src="https://github.com/skiptools/skip/assets/8979400/8ef50537-e39c-4903-8cd1-697fd5bf6946">

**Reply by joshuakcockrell** [reply]
Unfortunately, I'm not sure how to actually use the filter code I've written in Kotlin because it looks like `Just` doesn't work in Skip.

<img width="550" alt="Screenshot 2024-06-12 at 2 13 53 PM" src="https://github.com/skiptools/skip/assets/8979400/7a273cff-2ddf-4665-9eef-7a4a323da06f">

I really don't care if Skip supports `Just` or not (this is the only place in my entire app I use `Just`).

I just want _one_ way to solve this problem of users typing and copy pasting dumb strings.

**Reply by joshuakcockrell** [reply]
I even tried this approach (no `Just` required), which works in iOS, but not in Kotlin for some reason.

<img width="435" alt="Screenshot 2024-06-12 at 2 42 08 PM" src="https://github.com/skiptools/skip/assets/8979400/c42f77bb-8290-4078-935e-cbe78e04c298">

**Reply by joshuakcockrell** [reply]
[I'm just going through answers from this SO post](https://stackoverflow.com/a/58736068/4975772), trying to see what will work with Skip

This one looks like it would work 

`TextField("Placeholder", value: $text, format: .number)`

But it doesn't actually stop bad input on iOS.

**Reply by joshuakcockrell** [reply]
I feel like this attempt was promising. But it doesn't allow any input on Android, and the iOS version doesn't stop any bad characters
<img width="389" alt="Screenshot 2024-06-12 at 4 55 17 PM" src="https://github.com/skiptools/skip/assets/8979400/ef0e8211-94eb-4c66-b5ba-fd53efe58ab4">

**Reply by aabewhite** [reply]
Thanks for your patience! I'm going to look into this. In particular I'm not sure why the .onChange(...) thing isn't working.

In general, though, when you find some pretty simple Swift API that Skip doesn't support (like Set(_: String)), you might really consider adding support yourself through our open source libraries. It often isn't hard. The contributing section of the docs talks about how to get set up to contribute to the libraries:
https://skip.tools/docs/contributing/

**Comment by aabewhite** [comment]
I have this working so that you'll be able to do the onChange thing:

```
let allowedStrings = Set("...")

TextField(...).onChange(of: text) { _ in text = text.filter { allowedStrings.contains(text) } }
```

I'll let you know when the new releases are tagged so you can update your packages!

**Reply by aabewhite** [reply]
The releases have been tagged. If you File->Packages->Update to Latest, things should work as expected

**Reply by joshuakcockrell** [reply, marked as answer]
The release is working great! Thanks for the help.

Just for future reference if anyone sees this in the future, here's the final version that works on both iOS & Android.

<img width="560" alt="Screenshot 2024-06-18 at 9 12 07 AM" src="https://github.com/skiptools/skip/assets/8979400/1850b1ec-66c7-48f7-8b27-625197226d26">

```
func onlyDigits(_ input: String) -> String {
  #if SKIP
    let validChars = "0123456789".toSet() // Remove once Skip supports Set
  #else
    let validChars = Set("0123456789")
  #endif
  
  let out = input.filter{ validChars.contains($0) }
  print("in: \(input) out: \(out)")
  return out
}

public var body: some View {
  #if !SKIP
  // For some reason this is a bit glitchy on Android
    TextField("Placeholder", text: $text)
      .onChange(of: text) {_ in
        text = onlyDigits(text)
      }
  #else
    // For some reason this doesn't work on iOS
    TextField("Placeholder", text: .init(get: {
      return onlyDigits(text)
    }, set: { newValue in
      text = onlyDigits(newValue)
    }))
  #endif
}
  ```

**Reply by aabewhite** [reply]
Note that Set(String) should also work now
